### DISCUSSION TITLE
Offline mode not working

### DISCUSSION THREAD

**Discussion by nachodeh** [serwist/serwist#205] [Q&A]
My NextJS app is set up to serve a PWA with Serwist. I've followed the setup instructions and used the default config. The only change I added was `cacheOnNavigation: true`. On both Android and Mac OS (running on Chrome), I have the following issue:

1. Install PWA
2. Launch PWA with Internet enabled
3. App loads properly
4. Enable Airplane mode while app open
5. App immediately shows a "This site can't be reached" error
6. Console logs show https://react.dev/errors/418

If I add the page page as a fallback and additionalPrecacheEntry, then it works but is very clunky: reloads randomly, requires 2 clicks instead of 1 for each action, etc. If I only add it to additionalPrecacheEntries, it doesn't work.

Has anyone experienced this issue before? If the cacheOnNavigation is set to true and I'm currently on that page, what could cause the fetch to fail? My whole setup details are below.

cc @DuCanhGH since you seem to be the most active responder in this community :)

Deps:
```
"next": "^14.2.13",
"@serwist/next": "^9.0.8",
"serwist": "^9.0.8",
```

next.config.js
```
const { withContentlayer } = require("next-contentlayer")

/** @type {(phase: string, defaultConfig: import("next").NextConfig) => Promise<import("next").NextConfig>} */
module.exports = async (phase) => {
    /** @type {import("next").NextConfig} */
    const nextConfig = {
        images: {
            // dangerouslyAllowSVG: true,
            contentDispositionType: "attachment",
            remotePatterns: [
                {
                    protocol: "https",
                    hostname: "images.unsplash.com",
                },
                {
                    protocol: "https",
                    hostname: "api.producthunt.com",
                },
            ],
        },
        reactStrictMode: true,
        swcMinify: true,
    }

    const withSerwist = (await import("@serwist/next")).default({
        swSrc: "app/sw.ts",
        swDest: "public/sw.js",
        cacheOnNavigation: true,
    })

    return withSerwist(withContentlayer(nextConfig))
}
```

app/sw.js
```
import { defaultCache } from "@serwist/next/worker"
import type { PrecacheEntry, SerwistGlobalConfig } from "serwist"
import { Serwist } from "serwist"

// This declares the value of `injectionPoint` to TypeScript.
// `injectionPoint` is the string that will be replaced by the
// actual precache manifest. By default, this string is set to
// `"self.__SW_MANIFEST"`.
declare global {
    interface WorkerGlobalScope extends SerwistGlobalConfig {
        __SW_MANIFEST: (PrecacheEntry | string)[] | undefined
    }
}

declare const self: ServiceWorkerGlobalScope

const serwist = new Serwist({
    precacheEntries: self.__SW_MANIFEST,
    skipWaiting: true,
    clientsClaim: true,
    navigationPreload: true,
    runtimeCaching: defaultCache,
})

serwist.addEventListeners()
```

manifest.ts
```
import type { MetadataRoute } from "next"

export default function manifest(): MetadataRoute.Manifest {
    return {
        name: "appname",
        short_name: "appname",
        icons: [
            {
                src: "/logo-maskable.png",
                sizes: "512x512",
                type: "image/png",
                purpose: "maskable",
            },
            {
                src: "/logo.png",
                sizes: "512x512",
                type: "image/png",
            },
        ],
        theme_color: "#080E20",
        background_color: "#9333EA",
        start_url: "/dashboard",
        display: "standalone",
        orientation: "portrait",
        scope: "/",
    }
}
```

**Comment by DuCanhGH** [comment]
@nachodeh 

To avoid this issue, cache the pages ahead of time:

```ts
import { defaultCache } from "@serwist/next/worker";
import { type PrecacheEntry, Serwist } from "serwist";

declare global {
  interface WorkerGlobalScope {
    __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
  }
}

declare const self: ServiceWorkerGlobalScope;

const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST,
  precacheOptions: {
    cleanupOutdatedCaches: true,
    concurrency: 20,
  },
  runtimeCaching: defaultCache,
});

const urlsToCache = ["/", "/~offline"] as const;

self.addEventListener("install", (event) => {
  event.waitUntil(
    Promise.all(
      urlsToCache.map((entry) => {
        const request = serwist.handleRequest({
          request: new Request(entry),
          event,
        });
        return request;
      }),
    ),
  );
});

serwist.addEventListeners();
```

Note that this is a bit different from precaching. Precaching prevents your pages from being matched against any strategy other than the pre-defined `PrecacheStrategy`, whereas caching ahead of time allows just that.

Still, after the service worker has been installed, you will immediately see a working offline mode :)

> cc DuCanhGH since you seem to be the most active responder in this community :)

I am pretty much the only maintainer here :) However, I've been a bit too busy recently, so you may see some delay before I respond.

**Reply by xdmx** [reply]
> To avoid this issue, cache the pages ahead of time:
> 
> ```javascript
> const urlsToCache = ["/", "/~offline"] as const;
> 
> self.addEventListener("install", (event) => {
>   event.waitUntil(
>     Promise.all(
>       urlsToCache.map((entry) => {
>         const request = serwist.handleRequest({
>           request: new Request(entry),
>           event,
>         });
>         return request;
>       }),
>     ),
>   );
> });
> 
> serwist.addEventListeners();
> ```


I'm a bit curious about this approach.

I have a nextjs app and I'm trying to cache the homepage in order to have it working even when offline. So I think this might work, right? How would serwist take case of new updates on that page?

I tried a long time ago with a specific argument (don't remember which one, I think it got deprecated in v9, or maybe it was `additionalPrecacheEntries`, I'm not sure) and the problem I had was that the page was cached and working offline, but it'd never update when I made changes on that page (a part from hard refreshes, but users wouldn't do that)

**Reply by DuCanhGH** [reply]
@xdmx 

> and the problem I had was that the page was cached and working offline, but it'd never update when I made changes on that page (a part from hard refreshes, but users wouldn't do that)

That's what this fixes :) Unlike precaching, doing this allows the pages to be handled by the `runtimeCaching` strategies rather than `PrecacheStrategy`, which always uses precached files to handle requests.

**Comment by nachodeh** [comment]
Thanks @DuCanhGH !

I tried your approach and modified my sw.js, uninstalled and reinstalled the app, and the issue is still there.

After launching the app once with internet enabled, if I quit and restart the app in airplane mode, the error "This site can't be reached" shows up instead of the app.

This is my updated **sw.js**
```
import { defaultCache } from "@serwist/next/worker"
import type { PrecacheEntry, SerwistGlobalConfig } from "serwist"
import { Serwist } from "serwist"

// This declares the value of `injectionPoint` to TypeScript.
// `injectionPoint` is the string that will be replaced by the
// actual precache manifest. By default, this string is set to
// `"self.__SW_MANIFEST"`.
declare global {
    interface WorkerGlobalScope extends SerwistGlobalConfig {
        __SW_MANIFEST: (PrecacheEntry | string)[] | undefined
    }
}

declare const self: ServiceWorkerGlobalScope

const serwist = new Serwist({
    precacheEntries: self.__SW_MANIFEST,
    precacheOptions: {
      cleanupOutdatedCaches: true,
      concurrency: 20,
    },
    runtimeCaching: defaultCache,
  });
  
  const urlsToCache = ["/", "/dashboard"] as const;
  
  self.addEventListener("install", (event) => {
    event.waitUntil(
      Promise.all(
        urlsToCache.map((entry) => {
          const request = serwist.handleRequest({
            request: new Request(entry),
            event,
          });
          return request;
        }),
      ),
    );
  });
  
  serwist.addEventListeners();
```

**Comment by DuCanhGH** [comment]
@nachodeh you seem to have disabled `skipWaiting` and `clientsClaim`. Are you sure you had quit all open tabs before testing?

Ah, the example I provided did not include them. Please add them back if you were using them :) Also don't set `precacheOptions.concurrency` to `20` if you don't plan to precache every 20 entries concurrently. This is an optimization I added to the service worker of Serwist's documentation site since it has a lot to precache.

**Comment by nachodeh** [comment]
Hey @DuCanhGH, I tried that and initially it didn't work, so I tried to simplify my Component to dig into it. I found something quite interesting.

The problem seems to be with `useRouter from next/navigation`. I use it to update search params when the user navigates within screens of the app:

```
    const updateSearchParams = useCallback(
        (updates: { key: DashboardParams; value: string | null }[]) => {
            const params = new URLSearchParams()
            updates.forEach(({ key, value }) => {
                if (value) {
                    params.set(key, value)
                }
            })
            const newParams = params.toString()
            router.push(`?${newParams}`)
        },
        [router],
    )
```

When the hook that updates params is enabled with the router, I start seeing the errors. If I replace the router.push with the following, offline works:
```
const newParams = params.toString()
window.history.pushState({}, "", `?${newParams}`)
```

This is despite adding `precacheOptions.ignoreURLParametersMatching = [/.*/]`

```
const serwist = new Serwist({
    precacheEntries: self.__SW_MANIFEST,
    skipWaiting: true,
    clientsClaim: true,
    navigationPreload: true,
    precacheOptions: {
        cleanupOutdatedCaches: true,
        ignoreURLParametersMatching: [/.*/],
    },
    runtimeCaching: defaultCache,
})
```

Is this a known issue with `next/navigation`?

**Comment by DuCanhGH** [comment]
@nachodeh do this:

```ts
import { defaultCache, PAGES_CACHE_NAME } from "@serwist/next/worker";
import type { PrecacheEntry, SerwistGlobalConfig } from "serwist";
import { Serwist, NetworkFirst, ExpirationPlugin } from "serwist";

declare global {
  interface WorkerGlobalScope extends SerwistGlobalConfig {
    // Change this attribute's name to your `injectionPoint`.
    // `injectionPoint` is an InjectManifest option.
    // See https://serwist.pages.dev/docs/build/configuring
    __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
  }
}

declare const self: ServiceWorkerGlobalScope;

const matchOptions = {
  ignoreSearch: true,
} satisfies CacheQueryOptions;

const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: [
    {
      matcher: ({ request, url: { pathname }, sameOrigin }) =>
        request.headers.get("RSC") === "1" && request.headers.get("Next-Router-Prefetch") === "1" && sameOrigin && !pathname.startsWith("/api/"),
      handler: new NetworkFirst({
        cacheName: PAGES_CACHE_NAME.rscPrefetch,
        plugins: [
          new ExpirationPlugin({
            maxEntries: 32,
            maxAgeSeconds: 24 * 60 * 60, // 24 hours
          }),
        ],
        matchOptions,
      }),
    },
    {
      matcher: ({ request, url: { pathname }, sameOrigin }) => request.headers.get("RSC") === "1" && sameOrigin && !pathname.startsWith("/api/"),
      handler: new NetworkFirst({
        cacheName: PAGES_CACHE_NAME.rsc,
        plugins: [
          new ExpirationPlugin({
            maxEntries: 32,
            maxAgeSeconds: 24 * 60 * 60, // 24 hours
          }),
        ],
        matchOptions,
      }),
    },
    {
      matcher: ({ request, url: { pathname }, sameOrigin }) =>
        request.headers.get("Content-Type")?.includes("text/html") && sameOrigin && !pathname.startsWith("/api/"),
      handler: new NetworkFirst({
        cacheName: PAGES_CACHE_NAME.html,
        plugins: [
          new ExpirationPlugin({
            maxEntries: 32,
            maxAgeSeconds: 24 * 60 * 60, // 24 hours
          }),
        ],
        matchOptions,
      }),
    },
    ...defaultCache,
  ],
});

const urlsToCache = ["/", "/dashboard"] as const;

self.addEventListener("install", (event) => {
  event.waitUntil(
    Promise.all(
      urlsToCache.map((entry) => {
        const request = serwist.handleRequest({
          request: new Request(entry),
          event,
        });
        return request;
      }),
    ),
  );
});

serwist.addEventListeners();
```

Setting `matchOptions.ignoreSearch` to `true` tells `caches.match` to ignore search params.

**Reply by nachodeh** [reply]
Thanks for the help and support! Unfortunately, that didn't work either :(

**Comment by nachodeh** [comment]
In the end, I opted for moving away from `next/navigation` router and use `window.history.pushState()` instead. With this setup, the PWA works well in offline mode. It only errors when I try to refresh a page with params showing an error page instead of the offline page. It's good enough, though :)

Final config in case someone else is facing the same issue:

**`sw.js`**
```
import { defaultCache } from "@serwist/next/worker"
import type { PrecacheEntry, SerwistGlobalConfig } from "serwist"
import { Serwist } from "serwist"

// This declares the value of `injectionPoint` to TypeScript.
// `injectionPoint` is the string that will be replaced by the
// actual precache manifest. By default, this string is set to
// `"self.__SW_MANIFEST"`.
declare global {
    interface WorkerGlobalScope extends SerwistGlobalConfig {
        __SW_MANIFEST: (PrecacheEntry | string)[] | undefined
    }
}

declare const self: ServiceWorkerGlobalScope

const serwist = new Serwist({
    precacheEntries: self.__SW_MANIFEST,
    skipWaiting: true,
    clientsClaim: true,
    navigationPreload: true,
    disableDevLogs: true,
    precacheOptions: {
        cleanupOutdatedCaches: true,
        ignoreURLParametersMatching: [/.*/],
    },
    fallbacks: {
        entries: [
            {
                url: "/~offline",
                matcher({ request }) {
                    return request.destination === "document"
                },
            },
        ],
    },
    runtimeCaching: defaultCache,
})

const urlsToCache = ["/", "/dashboard", "/~offline"] as const

self.addEventListener("install", (event) => {
    event.waitUntil(
        Promise.all(
            urlsToCache.map((entry) => {
                const request = serwist.handleRequest({
                    request: new Request(entry),
                    event,
                })
                return request
            }),
        ),
    )
})

serwist.addEventListeners()
```

**`next.config.js`**
```
    const withSerwist = (await import("@serwist/next")).default({
        swSrc: "app/sw.ts",
        swDest: "public/sw.js",
        cacheOnNavigation: true,
        additionalPrecacheEntries: [{ url: "/~offline", revision: revision }],
    })
```

**Reply by DuCanhGH** [reply]
@nachodeh it took me a month to notice this comment :( Wonder what's happening inside `next/navigation` for this to happen, though.

**Reply by nachodeh** [reply]
No worries! In the end, it's mostly working so it's not a big deal. `next/navigation` is a pain to understand and debug :(

**Reply by mi-hol** [reply]
@nachodeh what version of next did you use?
I planned to actually create a doc site that must work offline using nextra V4 (see [here for more](https://github.com/shuding/nextra/discussions/2625)) now your experience is not that encouraging :(

**Reply by nachodeh** [reply]
> @nachodeh what version of next did you use?
> I planned to actually create a doc site that must work offline using nextra V4 (see [here for more](https://github.com/shuding/nextra/discussions/2625)) now your experience is not that encouraging :(

@mi-hol I'm using NextJS 14

**Reply by mapszk** [reply]
How can I cache routes with dynamic params for example /product/:id ? Is this possible?

**Reply by DuCanhGH** [reply]
@mapszk unfortunately not available automagically, as there's no way for Serwist to know what pages are prerendered. This is due to the way `@serwist/next` works: it being a mere webpack plugin is run before the routes are prerendered. This is a caveat that I was discussing with the guys at Vercel, but there aren't anything to solve this yet. You can still list which pages you want to cache by the method provided above though!

**Comment by daveycodez** [comment]
Is the install listener the recommended way to get a PWA working in Airplane mode? Can we update the docs to reflect this. PWA works out of the box without Serwist... I'm confused why it isn't working offline by default when Serwist is added. I followed the instructions for Next.js guide, turn on airplane mode, and fetch errors can't load the PWA at all

**Reply by daveycodez** [reply]
I've added this

```ts
const urlsToCache = ["/~offline", "/", "/chat", "/todos", "/blog", "/auth/sign-in"]

self.addEventListener("install", (event) => {
    event.waitUntil(
        Promise.all(
            urlsToCache.map((entry) => {
                const request = serwist.handleRequest({
                    request: new Request(entry),
                    event,
                })
                return request
            }),
        ),
    )
})
```

But what happens if this grows? For example I will have 10 pages and 10 locales, so that's 100 pages to cache. Will this lag opening the PWA on first launch? Or is this an async operation that happens while the user is browsing

**Reply by DuCanhGH** [reply]
@daveycodez your user can use the app just fine, but the service worker is kept in the `installing` phase until the promise settles. Also if you're caching around 100 pages on install then I recommend caching them in groups of ten rather than with `Promise.all`. This keeps the service worker in the `installing` phase for longer, but it prevents a laggy network. To be honest though, in your case, I'd recommend setting `register` to `false` and doing this instead:

```tsx
"use client";
import { useEffect } from "react";

export const onLocaleSwitch = (locale: string) => {
  if ("serviceWorker" in navigator && window.serwist) {
    // Get a bunch of URLs to cache.
    const urlsToCache = [];
    // Send that list of URLs to your router in the service worker.
    window.serwist.messageSW({
      type: "CACHE_URLS",
      payload: { urlsToCache },
    });
  }
}

export const Pwa = () => {
  useEffect(() => {
    if ("serviceWorker" in navigator && window.serwist) {
      window.serwist.addEventListener("activated", (event) => {
        // Get a bunch of URLs to cache. Use the current locale to
        // avoid caching URLs in locales that the user will never use.
        const urlsToCache = [];
        // Send that list of URLs to your router in the service worker.
        window.serwist.messageSW({
          type: "CACHE_URLS",
          payload: { urlsToCache },
        });
      });
      void window.serwist.register();
    }
  }, []);

  return <></>;
}
```
This allows you to detect a locale before fetching URLs, saving resources on URLs with locales that the user never uses. It also allows the service worker to activate much sooner, as there's no promise caching a hundred URLs blocking it from doing so.

**Reply by daveycodez** [reply]
@DuCanhGH What exactly is the purpose of additionalPrecacheEntries if we need to cache URL's at runtime? Is it necessary to include both?

**Reply by DuCanhGH** [reply]
@daveycodez `additionalPrecacheEntries` is just an array of URLs appended to at the end of `precacheEntries` / `self.__SW_MANIFEST`. The difference between using `additionalPrecacheEntries` and caching at install time is that the former only recaches when the `revision` has changed and prefers using cached entries to fetching from the network, while the latter recaches on every service worker install and relies on the Strategy of the `runtimeCaching` entries matching your URLs.

**Comment by fathisiddiqi** [comment]
I've an issue when using offline mode with Serwist. After the internet switched to Airplane mode, NextJS found pages from the cache/service worker. But NextJS didn't find other assets, so the request fell back to the offline page even though I've visited that page.

Here is my config

```
import { defaultCache } from '@serwist/next/worker';
import type { PrecacheEntry, SerwistGlobalConfig } from 'serwist';
import { Serwist } from 'serwist';

// This declares the value of `injectionPoint` to TypeScript.
// `injectionPoint` is the string that will be replaced by the
// actual precache manifest. By default, this string is set to
// `"self.__SW_MANIFEST"`.
declare global {
  interface WorkerGlobalScope extends SerwistGlobalConfig {
    __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
  }
}

declare const self: ServiceWorkerGlobalScope;

const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST as PrecacheEntry[],
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  precacheOptions: {
    cleanupOutdatedCaches: true,
  },
  fallbacks: {
    entries: [
      {
        url: `~offline.html`,
        matcher({ request }: { request: Request }) {
          return request.destination === 'document';
        },
      },
    ],
  },
  runtimeCaching: defaultCache, // Include defaultCache for other assets,
});

const LOCALES = ['en', 'id'] as const;

// Precache localized root pages only; protected routes (e.g., '/dashboard')
// should rely on runtime caching since they may require auth and differ per user.
const urlsFromLocales = Array.from(LOCALES, (l) => `/${l}`);

const urlsToCache = ['/', ...urlsFromLocales];

self.addEventListener('install', (event) => {
  event.waitUntil(
    Promise.all(
      urlsToCache.map((entry) => {
        const request = serwist.handleRequest({
          request: new Request(entry),
          event,
        });
        return request;
      })
    )
  );
});

serwist.addEventListeners();
```

**Reply by DuCanhGH** [reply]
@fathisiddiqi that's weird. Usually assets failing to load wouldn't trigger the fallbacks. It's rather the pages failing to load, which indicate that maybe your cached pages weren't used at all. Besides, "~offline.html" seems weird for a Next.js app, I'd say. Usually you wouldn't use paths ending with ".html" right?

**Reply by fathisiddiqi** [reply]
Actually don't know why SW got an error when trying to find in the cache. I can only use the cache for 30 seconds; after that, the cache is gone, and NextJS tries to use other rsc.

This is my current flow of my nextjs to Serwist
```
The first company-profile request (200 OK from service worker) → ✅ this one is being served correctly from your pages-rsc cache.

The second company-profile request (red X, no response headers) → ❌ this one bypasses cache and goes directly as a document navigation (HTML request, not RSC).
```

Is the default runtime caching already caching the HTML request?

**Reply by fathisiddiqi** [reply]
I just put my offline fallback in the public directory as html file. I've tried to put in `app/~offline/page.tsx`, but when using precache entries, sw returns an error. Thus, offline fallback didn't cache.

**Reply by DuCanhGH** [reply]
@fathisiddiqi `defaultCache` would try to cache both HTML and RSC responses, yes. Usually the `pages` cache would be used when the page is loaded using a normal browser navigation (e.g. a page reload) or when `cacheOnNavigation` is enabled. The `pages-rsc` cache would be used when the page is loaded by Next.js (e.g. `next/link`). That the second request sends a HTML request is kinda weird, and you'd need a cached HTML response from `pages` to respond to such a request. I don't really think the cached `pages-rsc` response was thrown away by Serwist.

**Comment by Almuhanad-1** [comment]
i faced the same issue (offline page not working) and fixed by adding additionalPrecacheEntries to the withSerwistInit in next.config.ts
```
const withSerwist = withSerwistInit({
  swSrc: "src/app/sw.ts",
  swDest: "public/sw.js",
  cacheOnNavigation: true,
  reloadOnOnline: true,
  register: true, // Enable automatic registration
  additionalPrecacheEntries: [
    { url: "/en/offline", revision: crypto.randomUUID() },
    { url: "/ar/offline", revision: crypto.randomUUID() },
    { url: "/offline", revision: crypto.randomUUID() },
  ],
  disable: process.env.NODE_ENV !== "production", // Disable in development for easier debugging
});
```

i hope this helps anyone is facing the same issue
